x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/type/EsqlDataTypeConverterTests.java

```diff
@@ -30,23 +56,123 @@ public void testNanoTimeToString() {
+                } else if ((isString(dataType1) && isString(dataType2))) {
+                    if (dataType1 == dataType2) {
+                        assertEqualsCommonType(dataType1, dataType2, dataType1);
+                    } else {
+                        assertEqualsCommonType(dataType1, dataType2, TEXT);
+                    }
+                } else {
+                    assertNullCommonType(dataType1, dataType2);
+                }
+            }
+        }
+    }
 
-        // strings
-        assertEquals(TEXT, commonType(TEXT, KEYWORD));
-        assertEquals(TEXT, commonType(KEYWORD, TEXT));
+    public void testCommonTypeDateTimeIntervals() {
+        List<DataType> DATE_TIME_INTERVALS = List.of(DATETIME, DATE_PERIOD, TIME_DURATION);
+        for (DataType dataType1 : DATE_TIME_INTERVALS) {
+            for (DataType dataType2 : DataType.values()) {
+                if (dataType2 == NULL) {
+                    assertEqualsCommonType(dataType1, NULL, dataType1);
+                } else if (DATE_TIME_INTERVALS.containsAll(List.of(dataType1, dataType2))) {
+                    if (isDateTime(dataType1) || isDateTime(dataType2)) {
+                        assertEqualsCommonType(dataType1, dataType2, DATETIME);
+                    } else if (dataType1 == dataType2) {
+                        assertEqualsCommonType(dataType1, dataType2, dataType1);
+                    } else {
+                        assertNullCommonType(dataType1, dataType2);
+                    }
+                } else {
+                    assertNullCommonType(dataType1, dataType2);
+                }
+            }
+        }
     }
 
+    public void testCommonTypeNumeric() {
+        // whole numbers
+        commonNumericType(BYTE, List.of(NULL, BYTE));
+        commonNumericType(SHORT, List.of(NULL, BYTE, SHORT));
+        commonNumericType(INTEGER, List.of(NULL, BYTE, SHORT, INTEGER));
+        commonNumericType(LONG, List.of(NULL, BYTE, SHORT, INTEGER, LONG));
+        commonNumericType(UNSIGNED_LONG, List.of(NULL, BYTE, SHORT, INTEGER, LONG, UNSIGNED_LONG));
+        // floats
+        commonNumericType(HALF_FLOAT, List.of(NULL, BYTE, SHORT, INTEGER, LONG, UNSIGNED_LONG, HALF_FLOAT, FLOAT));
+        commonNumericType(FLOAT, List.of(NULL, BYTE, SHORT, INTEGER, LONG, UNSIGNED_LONG, FLOAT, HALF_FLOAT));
+        commonNumericType(DOUBLE, List.of(NULL, BYTE, SHORT, INTEGER, LONG, UNSIGNED_LONG, HALF_FLOAT, FLOAT, DOUBLE, SCALED_FLOAT));
+        commonNumericType(SCALED_FLOAT, List.of(NULL, BYTE, SHORT, INTEGER, LONG, UNSIGNED_LONG, HALF_FLOAT, FLOAT, SCALED_FLOAT, DOUBLE));
+    }
+
+    /**
+     * The first argument and the second argument(s) have the first argument as a common type.
+     */
+    private static void commonNumericType(DataType numericType, List<DataType> lowerTypes) {
+        List<DataType> NUMERICS = List.of(INTEGER, LONG, DOUBLE, UNSIGNED_LONG, BYTE, SHORT, FLOAT, HALF_FLOAT, SCALED_FLOAT);
```

Comment:
Same here about the list of numerics and doubles. Use helper methods from `DataType` (quick look at it reveals `isNumeric` and `isRationalNumber` but there may be others in there better suited).